aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -355,15 +346,18 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
-  Tensor ret = at::empty(self.sizes(), self.options());
+  auto result_type = self.device().is_cpu()
```

Comment:
This is going to need a comment (possibly a pointer to the where type promotion note referenced above).

Could this check be simplified to testing if self and other have distinct dtypes and, if they do, asserting they're not on the CPU?